### PR TITLE
Add consumer support for zosma

### PR DESCRIPTION
**Pull request by thumbnail**
As well as a short form, thanks @romnous

### CODE REVIEW

File: src/main/java/io/leonis/zosma/game/engine/Zosma.java [hunk 1]
```diff
@@ -1,38 +1,112 @@
...
+      final Function<Publisher<I>, Publisher<J>> inputAdapter,
+      final Function<Publisher<J>, Publisher<Q>> ai,
+      final Function<Publisher<Q>, Publisher<O>> outputAdapter,
+      final Subscriber<O> outputSubscriber
+  ) {
+    this(inputPublisher, inputAdapter.andThen(ai).andThen(outputAdapter), outputSubscriber);
+  }
+
+  /**
+   * Constructs a new game agent which contains adapters for dealing with external types.
```
Comment (romni) [inline src/main/java/io/leonis/zosma/game/engine/Zosma.java]: this is the constructor without adapters

File: src/main/java/io/leonis/zosma/game/engine/Zosma.java [hunk 2]
```diff
@@ -1,38 +1,112 @@
...
+      final Function<Publisher<I>, Publisher<J>> inputAdapter,
+      final Function<Publisher<J>, Publisher<Q>> ai,
+      final Function<Publisher<Q>, Publisher<O>> outputAdapter,
+      final Consumer<O> outputConsumer
+  ) {
+    this(inputPublisher, inputAdapter.andThen(ai).andThen(outputAdapter), outputConsumer);
+  }
+
+  /**
+   * Constructs a new game agent which contains adapters for dealing with external types.
```
Comment (romni) [inline src/main/java/io/leonis/zosma/game/engine/Zosma.java]: this constructor is also without adapters

File: src/main/java/io/leonis/zosma/game/engine/Zosma.java [hunk 3]
```diff
@@ -1,38 +1,112 @@
...
+   * @param <Q>              The type of output produced by the game agent.
+   */
+  public <I, J, Q, O> Zosma(
+      final Publisher<I> inputPublisher,
+      final Function<Publisher<I>, Publisher<J>> inputAdapter,
+      final Function<Publisher<J>, Publisher<Q>> ai,
+      final Function<Publisher<Q>, Publisher<O>> outputAdapter,
+      final Consumer<O> outputConsumer
+  ) {
+    this(inputPublisher, inputAdapter.andThen(ai).andThen(outputAdapter), outputConsumer);
```
Comment (romni) [inline src/main/java/io/leonis/zosma/game/engine/Zosma.java:86]: nice reversing that

**Review by romni — CHANGES_REQUESTED** [review]

**Comment by thumbnail** [comment]
I've had another thought; it's technically possible to use method references to use one primary constructor instead of two. the primary constructor would be `Zosma(Publisher<I>, Function<I, O>, Consumer<O>, Consumer<Throwable>, Runnable, Consumer<Subscription>)`. which is probably more confusing than anything else

``` java
  public <I, O> Zosma(
      final Publisher<I> inputPublisher,
      final Function<Publisher<I>, Publisher<O>> ai,
      final Subscriber<O> outputSubscriber
  ) {
   this(inputPublisher, ai, outputSubscriber::onNext, outputSubscriber::onError, outputSubscriber::onComplete, outputSubscriber::onSubscribe)
 }
```

**Review by romni — APPROVED** [review]

**Pull request merged**